File: agent/local_sales.py
"""
Ventas/Facturas locales por perfil.
"""
import json
from pathlib import Path
from agent.profiles import get_profile_data_dir


def _sales_file(profile_id: str) -> Path:
    return get_profile_data_dir(profile_id) / "sales.json"


def _load(profile_id: str) -> list:
    f = _sales_file(profile_id)
    return json.loads(f.read_text(encoding="utf-8")) if f.exists() else []


def _save(profile_id: str, records: list):
    _sales_file(profile_id).write_text(json.dumps(records, ensure_ascii=False, indent=2), encoding="utf-8")


def record_invoice(profile_id: str, invoice: dict):
    records = _load(profile_id)
    records.append(invoice)
    _save(profile_id, records)


def list_invoices(profile_id: str, limit: int = 200) -> list[dict]:
    records = _load(profile_id)
    return sorted(records, key=lambda x: x.get("timestamp", ""), reverse=True)[:limit]


def delete_record(profile_id: str, sale_id: str):
    records = [r for r in _load(profile_id)
               if r.get("sale_id") != sale_id and r.get("invoice_id") != sale_id]
    _save(profile_id, records)

File: desktop_java/src/main/java/com/mfagent/service/LocalSales.java
package com.mfagent.service;

import com.mfagent.model.Invoice;

import java.nio.file.Path;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

/**
 * CRUD de ventas/facturas locales por perfil.
 * Equivalente a agent/local_sales.py
 */
public class LocalSales {

    private final Path file;

    public LocalSales(String profileId) {
        this.file = ProfilePaths.dataDir(profileId).resolve("sales.json");
    }

    public List<Invoice> list(int limit) {
        return JsonStore.readList(file, Invoice.class)
                .stream()
                .sorted(Comparator.comparing(
                        (Invoice i) -> i.getTimestamp() != null ? i.getTimestamp() : "")
                        .reversed())
                .limit(limit)
                .collect(Collectors.toList());
    }

    public List<Invoice> list() {
        return list(200);
    }

    public void record(Invoice invoice) {
        List<Invoice> records = JsonStore.readList(file, Invoice.class);
        // Evitar duplicados por invoice_id
        String id = invoice.getInvoiceId();
        if (id != null && records.stream().anyMatch(r -> id.equals(r.getInvoiceId()))) {
            return;
        }
        records.add(invoice);
        JsonStore.writeList(file, records);
    }

    public void delete(String invoiceId) {
        if (invoiceId == null) return;
        List<Invoice> records = JsonStore.readList(file, Invoice.class);
        records.removeIf(r -> invoiceId.equals(r.getInvoiceId()));
        JsonStore.writeList(file, records);
    }

    /** Borra por referencia de objeto (para facturas sin invoiceId). */
    public void deleteByRef(Invoice target) {
        List<Invoice> records = JsonStore.readList(file, Invoice.class);
        // Identificar por timestamp + customer + total
        records.removeIf(r ->
            java.util.Objects.equals(r.getTimestamp(), target.getTimestamp()) &&
            java.util.Objects.equals(r.getCustomer(),  target.getCustomer())  &&
            r.getTotal() == target.getTotal()
        );
        JsonStore.writeList(file, records);
    }

    public void update(String invoiceId, Invoice updated) {
        List<Invoice> records = JsonStore.readList(file, Invoice.class);
        for (int i = 0; i < records.size(); i++) {
            if (invoiceId.equals(records.get(i).getInvoiceId())) {
                records.set(i, updated);
                break;
            }
        }
        JsonStore.writeList(file, records);
    }
}
